sct-commons/src/main/java/org/lfenergy/compas/sct/commons/scl/ied/AbstractLNAdapter.java

```diff
@@ -360,50 +344,51 @@ public List<ControlBlock<?>> getControlBlocksForMatchingFCDA(@NonNull ExtRefInfo
         List<ControlBlock<?>> controlBlocks = new ArrayList<>();
-        List<?> tControls;
-
-        LNodeMetaData metaData = LNodeMetaData.from(this);
-
         for (TDataSet tDataSet : tDataSets) {
-            if (isLN0() && (serviceType == null || serviceType == TServiceType.GOOSE)) {
-                tControls = this.lookUpControlBlocksByDataSetRef(tDataSet.getName(), TGSEControl.class);
-                controlBlocks.addAll(
-                        tControls.stream()
-                                .map(tgseControl -> {
-                                    var g = new GooseControlBlock((TGSEControl) tgseControl);
-                                    g.setMetaData(metaData);
-                                    return g;
-                                })
-                                .collect(Collectors.toList())
-                );
-            }
-
-            if (isLN0() && (serviceType == null || serviceType == TServiceType.SMV)) {
-
-                tControls = this.lookUpControlBlocksByDataSetRef(tDataSet.getName(), TSampledValueControl.class);
-                controlBlocks.addAll(
-                        tControls.stream()
-                                .map(sampledValueControl -> {
-                                    var s = new SMVControlBlock((TSampledValueControl) sampledValueControl);
-                                    s.setMetaData(metaData);
-                                    return s;
-                                })
-                                .collect(Collectors.toList())
-                );
-            }
+            controlBlocks.addAll(getControlBlocksByDataSetRef(tDataSet.getName(), serviceType));
+        }
+        return controlBlocks;
+    }
 
-            if (serviceType == null || serviceType == TServiceType.REPORT) {
-                tControls = this.lookUpControlBlocksByDataSetRef(tDataSet.getName(), TReportControl.class);
-                controlBlocks.addAll(
-                        tControls.stream()
-                                .map(reportControl -> {
-                                    var r = new ReportControlBlock((TReportControl) reportControl);
-                                    r.setMetaData(metaData);
-                                    return r;
-                                })
-                                .collect(Collectors.toList())
-                );
-            }
+    /**
+     * finds all Control Blocks by dataSetRef and a Service Type (GOOSE, SampleValue, Report) or all Service Types.
+     * @param dataSetRef DatSet value searched
+     * @param serviceType service type to be filtered
+     * @return all Control Blocks matching dataSetRef and a Service Type or all Service Types
+     */
+    private List<ControlBlock<?>> getControlBlocksByDataSetRef(String dataSetRef, TServiceType serviceType) {
+        List<ControlBlock<?>> controlBlocks = new ArrayList<>();
```

Comment:
You can work with streams and use "concat" to concat them together

Reply:
ok